Show files produced by `--emit foo` in json artifact notifications

pacak opened the pull request:
Right now it is possible to ask `rustc` to save some intermediate representation into one or more files with `--emit=foo`, but figuring out what exactly was produced is difficult. This pull request adds information about `llvm_ir` and `asm` intermediate files into notifications produced by `--json=artifacts`.

Related discussion: https://internals.rust-lang.org/t/easier-access-to-files-generated-by-emit-foo/20477

Motivation - `cargo-show-asm` parses those intermediate files and presents them in a user friendly way, but right now I have to apply some dirty hacks. Hacks make behavior confusing: https://github.com/hintron/computer-enhance/issues/35

This pull request introduces a new behavior: now `rustc` will emit a new artifact notification for every artifact type user asked to `--emit`, for example for `--emit asm` those will include all the `.s` files.

Most users won't notice this behavior, to be affected by it all of the following must hold:
- user must use `rustc` binary directly (when `cargo` invokes `rustc` - it consumes artifact notifications and doesn't emit anything)
- user must specify both `--emit xxx` and `--json artifacts`
- user must refuse to handle unknown artifact types
- user must disable incremental compilation (or deal with it better than cargo does, or use a workaround like `save-temps`) in order not to hit #88829 / #89149

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8308528341/job/22738862706) [(plain)](https://github.com/rust-lang/rust/commit/0ded519383e2b2ad38fea812c2884a963ed17044/checks/22738862706/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking tempfile v3.7.1
error[E0308]: mismatched types
    --> src/back/write.rs:93:48
     |
93   |                 dcx.emit_artifact_notification(path, "asm");
     |                     -------------------------- ^^^^ expected `&Path`, found `PathBuf`
     |                     arguments to this method are incorrect
     |
note: method defined here
    --> /checkout/compiler/rustc_errors/src/lib.rs:1007:12
    --> /checkout/compiler/rustc_errors/src/lib.rs:1007:12
     |
1007 |     pub fn emit_artifact_notification(&self, path: &Path, artifact_type: &str) {
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider borrowing here
     |
93   |                 dcx.emit_artifact_notification(&path, "asm");

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
Build completed unsuccessfully in 0:02:25
```

</details>

petrochenkov commented:
Seems reasonable.
r? @bjorn3 do you have an opinion on this?

bjorn3 commented on compiler/rustc_codegen_gcc/src/back/write.rs:
```diff
@@ -78,14 +78,20 @@ pub(crate) unsafe fn codegen(
-            std::fs::write(out, "").expect("write file");
+            std::fs::write(&out, "").expect("write file");
+            if config.json_artifact_notifications {
+                dcx.emit_artifact_notification(&out, "llvm_ir");
```
Are there any tools which depend on artifact notifications of a given type only occurring once?

pacak replied:
I'm not aware of such tools. I did check `cargo-metadata` before making this change and it seem to simply copy list of files as is.

pacak replied:
Correction, `cargo` itself consumes this list and ignores it: https://github.com/rust-lang/cargo/issues/13672, `cargo-metadata` simply passes what it gets from cargo.

So to be affected you need to use `rustc` directly, ask it to emit artifact notifications, ask it to emit llvm/asm/etc and fail if previously unknown artifact type occurs, optionally more than once.

bjorn3 commented on src/doc/rustc/src/json.md:
```diff
@@ -229,6 +229,8 @@ flag][option-emit] documentation.
        - "dep-info": The `.d` file with dependency information in a Makefile-like syntax.
        - "metadata": The Rust `.rmeta` file containing metadata about the crate.
+       - "asm": The `.s` file with generated assembly
+       - "llvm_ir": The `.ll` file with generated textual LLVM IR
```
For symmetry adding the `obj` kind would make sense to me.

bjorn3 replied:
Also the current description reads like there can be only one file of the asm and llvm_ir kinds.

pacak replied:
> For symmetry adding the obj kind would make sense to me.

Sure, will look into adding that as well as all other file types listed [here](https://doc.rust-lang.org/rustc/command-line-arguments.html#--emit-specifies-the-types-of-output-files-to-generate)

pacak commented:
I changed the documentation a bit, now it mentions that notifications can have file of the same type mentioned several times, added support for all the files that can be emitted (I hope), changed `llvm_bc` -> `llvm-bc` to match the documentation.

Now I'm getting something like this:

```
% /home/pacak/ej/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc --crate-name smo --edition=2021 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=239 --crate-type lib --emit=dep-info,metadata,link,llvm-ir,llvm-bc,asm,mir,obj -C opt-level=3 -C codegen-units=16
{"$message_type":"artifact","artifact":"smo.d","emit":"dep-info"}
{"$message_type":"artifact","artifact":"libsmo.rmeta","emit":"metadata"}
{"$message_type":"artifact","artifact":"smo.mir","emit":"mir"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.2.rcgu.bc","emit":"llvm-bc"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.2.rcgu.ll","emit":"llvm-ir"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.0.rcgu.bc","emit":"llvm-bc"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.0.rcgu.ll","emit":"llvm-ir"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.2.rcgu.s","emit":"asm"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.2.rcgu.o","emit":"obj"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.1.rcgu.bc","emit":"llvm-bc"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.1.rcgu.ll","emit":"llvm-ir"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.0.rcgu.s","emit":"asm"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.0.rcgu.o","emit":"obj"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.1.rcgu.s","emit":"asm"}
{"$message_type":"artifact","artifact":"smo.smo.9eb4b431c6ed4cc9-cgu.1.rcgu.o","emit":"obj"}
{"$message_type":"diagnostic","message":"ignoring emit path because multiple .bc files were produced","code":null,"level":"warning","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[33mwarning\u001b[0m\u001b[0m\u001b[1m: ignoring emit path because multiple .bc files were produced\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"ignoring emit path because multiple .s files were produced","code":null,"level":"warning","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[33mwarning\u001b[0m\u001b[0m\u001b[1m: ignoring emit path because multiple .s files were produced\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"ignoring emit path because multiple .ll files were produced","code":null,"level":"warning","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[33mwarning\u001b[0m\u001b[0m\u001b[1m: ignoring emit path because multiple .ll files were produced\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"ignoring emit path because multiple .o files were produced","code":null,"level":"warning","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[33mwarning\u001b[0m\u001b[0m\u001b[1m: ignoring emit path because multiple .o files were produced\u001b[0m\n\n"}
{"$message_type":"artifact","artifact":"libsmo.rlib","emit":"link"}
{"$message_type":"diagnostic","message":"4 warnings emitted","code":null,"level":"warning","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[33mwarning\u001b[0m\u001b[0m\u001b[1m: 4 warnings emitted\u001b[0m\n\n"}
```

Error messages are wrong - I'm not passing an emit path, ticket is here https://github.com/rust-lang/rust/issues/122509 - I think I'll just make a fix later.

Now I'm trying to figure out how to run parser from cargo-metadata on this output...

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8376627726/job/22936762887) [(plain)](https://github.com/rust-lang/rust/commit/50538e5e2601607f4ca21c80caaaf39a2c9d6985/checks/22936762887/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 46.4s done
#16 DONE 57.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
##[group]Clock drift check
  local time: Thu Mar 21 14:27:58 UTC 2024
  network time: Thu, 21 Mar 2024 14:27:58 GMT
  network time: Thu, 21 Mar 2024 14:27:58 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/json/json-multiple.rs stdout ----
diff of stderr:

+ {"$message_type":"artifact","artifact":"$TEST_BUILD_DIR/json/json-multiple/json-multiple.json_multiple.8dc5531ec5ab66e7-cgu.0.rcgu.o","emit":"obj"}
1 {"$message_type":"artifact","artifact":"$TEST_BUILD_DIR/json/json-multiple/libjson_multiple.rlib","emit":"link"}


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/json/json-multiple/json-multiple.stderr
---

---- [ui] tests/ui/json/json-options.rs stdout ----
diff of stderr:

+ {"$message_type":"artifact","artifact":"$TEST_BUILD_DIR/json/json-options/json-options.json_options.a0c316e8fd97a360-cgu.0.rcgu.o","emit":"obj"}
1 {"$message_type":"artifact","artifact":"$TEST_BUILD_DIR/json/json-options/libjson_options.rlib","emit":"link"}


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/json/json-options/json-options.stderr
```

</details>

pacak commented:
Well, failing tests are asking to emit obj files, directly or indirectly as well as json information about artifacts. test runner finds unexpected new information about the obj files and fails. This new information is kind of the whole point of this pull request so saving this as expected output makes sense, but I'm not sure about how stable random looking parts of the name are - I don't want to introduce flaky tests. 

I think my options are
- don't emit information about obj files unless --emit obj is present
- figure out how to normalize random looking part of the name and save new expected test output...

Thoughts?

bjorn3 commented on compiler/rustc_codegen_gcc/src/back/write.rs:
```diff
@@ -113,6 +125,9 @@ pub(crate) unsafe fn codegen(
                 }
+                if config.json_artifact_notifications {
+                    dcx.emit_artifact_notification(&obj_out, "obj");
+                }
```
This should be moved after the `compile_to_file` calls. Otherwise the notification would be emitted without actually having written any object file.

bjorn3 commented on compiler/rustc_codegen_gcc/src/back/write.rs:
```diff
@@ -70,6 +73,9 @@ pub(crate) unsafe fn codegen(
                 context.add_command_line_option("-ffat-lto-objects");
+                if config.json_artifact_notifications {
+                    dcx.emit_artifact_notification(&bc_out, "llvm-bc");
+                }
```
Same

bjorn3 commented on compiler/rustc_codegen_gcc/src/back/write.rs:
```diff
@@ -56,6 +56,9 @@ pub(crate) unsafe fn codegen(
                 context.add_command_line_option("-flto-partition=one");
+                if config.json_artifact_notifications {
+                    dcx.emit_artifact_notification(&bc_out, "llvm-bc");
+                }
```
Same

bjorn3 reviewed (commented):
Would you mind also adding this to rustc_codegen_cranelift?

The changes in this PR LGTM minus the couple of small changes below. I don't think I can approve an insta-stable change like myself this though. Maybe put it behind an unstable flag?

pacak commented:
> Would you mind also adding this to rustc_codegen_cranelift?

Is it actually producing intermediate artifacts? I didn't add it because I couldn't find any apart from `crate::pretty_clif::write_clif_file` which feels like it's doing something slightly different. I can add support for it or any other places if you point them to me,


> The changes in this PR LGTM minus the couple of small changes below.

fixed

> I don't think I can approve an insta-stable change like myself this though. Maybe put it behind an unstable flag?

Hmm... Should we poke somebody else then? It's not really a behavior change, more like extending existing diagnostic that you have to ask specifically for with two separate flags - emit intermediates and emit artifact notifications.

bjorn3 commented:
> Hmm... Should we poke somebody else then?

You could ask on zulip in `#t-compiler`. Or maybe ask during a compiler team meeting: https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings (every thursday, not sure about what time it would be in your timezone)

pacak commented:
> Would you mind also adding this to rustc_codegen_cranelift?

I'm still not sure where it produces those object files or how to compile and run rustc with cranelift. If you can point me at exact line/function - I'll try to add the support, otherwise - I'll pass.

bjorn3 commented:
The object files for a cgu are all emitted right before https://github.com/rust-lang/rust/blob/fa374a80715253204225ce33b5e931aade9351c2/compiler/rustc_codegen_cranelift/src/driver/aot.rs#L175 It is probably easier to iterate through `modules` at https://github.com/rust-lang/rust/blob/fa374a80715253204225ce33b5e931aade9351c2/compiler/rustc_codegen_cranelift/src/driver/aot.rs#L108 though as you have access to the Session there and object files are copied from the incr comp cache just a bit earlier in this function. By the way are you emitting artifact notifications for object files copied from the incr comp cache for the LLVM backend?

pacak commented:
> The object files for a cgu are all emitted right before 

Okay, at least I was looking in the right file :) Added, seem to work.

> By the way are you emitting artifact notifications for object files copied from the incr comp cache for the LLVM backend?

Added this as well. Now I'm wondering how notifications for asm/llvm are interacting with incremental cache...

pacak commented:
I rebased on top of #123221, this allowed to simplify the logic a bunch, this pull request should be merged only after #123221 is. Did a bunch of tests locally, seem to be doing the right thing with one or multiple codegen units. I might add more tests later.

jieyouxu commented on tests/run-make/emit-notify/rmake.rs:
```diff
@@ -0,0 +1,37 @@
+    for _ in 0..=1 {
+        let output = rustc()
+            .input("lib.rs")
+            .arg("--emit=obj,asm,llvm-ir,llvm-bc,mir")
```
Could you introduce some semantically meaningful helpers like

```
rust()
    .emit("obj,asm,llvm-ir,...")
    .codegen_units(1)
    .json("artifacts")
    .error_format("json")
    .incremental(inc_dir) // incremental<P: AsRef<Path>>(path: P)
```

pacak replied:
> Could you introduce some semantically meaningful helpers like

Sure. Here or in a separate pull request?

jieyouxu replied:
> Sure. Here or in a separate pull request?

In this PR is fine, since you'll be using the helpers in this test

jieyouxu commented on src/tools/run-make-support/src/rustc.rs:
```diff
@@ -85,6 +92,35 @@ impl Rustc {
+    }
+
+    /// Specify error format to use
+    pub fn error_format(&mut self, format: ErrorFormat) -> &mut Self {
```
minor nit: for the API design, we experimented with it and ended up with a weakly-typed API with helper methods (see #122460), so this can just be

```rs
error_format(&mut self, format: &str)
```

jieyouxu commented on tests/run-make/asm-incr-cache/rmake.rs:
```diff
@@ -0,0 +1,16 @@
+    for _ in 0..=1 {
+        rustc()
+            .input("lib.rs")
+            .arg("--emit=obj,asm,dep-info,link,mir,llvm-ir,llvm-bc")
```
nit: you could also use the helpers you introduced for this test

pacak replied:
Technically this gets added in a different pull request so I just add a separate commit in this one where helpers are added to update the test.

jieyouxu commented on tests/run-make/emit-notify/lib.rs:
```diff
@@ -0,0 +1,23 @@
+//@ compile-flags: -C codegen-units=1 --emit asm,llvm-ir,llvm-bc,obj --json=artifacts --error-format=json
```
be advised that these compiletest flags **only** affect rmake.rs, not other Rust files, and `compile-flags` don't do anything for `rmake.rs`.

pacak replied:
Makes sense. It's a leftover from failed attempts to write this test without rmake :)

jieyouxu reviewed (commented):
Left some more minor nits :3

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8501722432/job/23284933655) [(plain)](https://github.com/rust-lang/rust/commit/78921016671570c592252e3f0c024f01ba905232/checks/23284933655/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.9s done
#16 DONE 33.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.32s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
failures:

---- [run-make] tests/run-make/emit-notify stdout ----

error: run-make test failed: could not build `rmake.rs` recipe
status: exit status: 1
command: env -u CARGO_MAKEFLAGS -u MAKEFLAGS -u MFLAGS HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerparallel dwp engine executionengine extensions filecheck frontendhlsl frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUST_BUILD_STAGE="stage2-x86_64-unknown-linux-gnu" S="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TMPDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-notify/emit-notify" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-notify/emit-notify/rmake" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--extern" "run_make_support=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/librun_make_support.rlib" "/checkout/tests/run-make/emit-notify/rmake.rs"
--- stderr -------------------------------
error[E0432]: unresolved import `run_make_support::ErrorFormat`
Build completed unsuccessfully in 0:36:29
##[error] --> /checkout/tests/run-make/emit-notify/rmake.rs:3:40
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8502063111/job/23285751954) [(plain)](https://github.com/rust-lang/rust/commit/dd1e1abfc7045f1f6997e20b6a0b19cea095c479/checks/23285751954/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.5s done
#17 DONE 35.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.43s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
failures:

---- [run-make] tests/run-make/asm-incr-cache stdout ----

error: run-make test failed: could not build `rmake.rs` recipe
status: exit status: 1
command: env -u CARGO_MAKEFLAGS -u MAKEFLAGS -u MFLAGS HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerparallel dwp engine executionengine extensions filecheck frontendhlsl frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUST_BUILD_STAGE="stage2-x86_64-unknown-linux-gnu" S="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TMPDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/asm-incr-cache/asm-incr-cache" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/asm-incr-cache/asm-incr-cache/rmake" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--extern" "run_make_support=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/librun_make_support.rlib" "/checkout/tests/run-make/asm-incr-cache/rmake.rs"
Build completed unsuccessfully in 0:37:05
--- stderr -------------------------------
error[E0382]: use of moved value: `inc_dir`
error[E0382]: use of moved value: `inc_dir`
##[error]  --> /checkout/tests/run-make/asm-incr-cache/rmake.rs:13:26
6  |     let inc_dir = tmp_dir();
6  |     let inc_dir = tmp_dir();
   |         ------- move occurs because `inc_dir` has type `PathBuf`, which does not implement the `Copy` trait
8  |     for _ in 0..=1 {
   |     -------------- inside of this loop
...
...
13 |             .incremental(inc_dir)
   |                          ^^^^^^^ value moved here, in previous iteration of loop
help: consider moving the expression out of the loop so it is only moved once
   |
8  ~     let mut value = rustc()
8  ~     let mut value = rustc()
9  +             .input("lib.rs")
10 +             .emit("obj,asm,dep-info,link,mir,llvm-ir,llvm-bc")
11 +             .arg("--crate-type=lib")
12 +             .incremental(inc_dir);
13 ~     for _ in 0..=1 {
   |
help: consider cloning the value if the performance cost is acceptable
   |
   |
13 |             .incremental(inc_dir.clone())

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0382`.
```

</details>

jieyouxu commented on tests/run-make/emit-notify/rmake.rs:
```diff
@@ -0,0 +1,37 @@
+extern crate run_make_support;
```
Same for this test.

pacak commented on tests/run-make/asm-incr-cache/rmake.rs:
```diff
@@ -0,0 +1,16 @@
+extern crate run_make_support;
```
Done. When making tests I just looked at existing tests, saw `Makefile`, lack of specialized helpers, lack of comments...

jieyouxu commented on src/tools/run-make-support/src/rustc.rs:
```diff
@@ -85,6 +85,31 @@ impl Rustc {
+
+    /// Specify directory path used for incremental cache
+    pub fn incremental(&mut self, path: impl AsRef<Path>) -> &mut Self {
+        let path = path.as_ref().to_string_lossy();
+        self.cmd.arg(format!("-Cincremental={}", path));
```
Could you avoid the lossy conversion and instead construct this argument as an `OsString`?

pacak committed:
Show files produced by --emit foo in json artifact notifications

pacak commented:
https://github.com/rust-lang/rust/pull/123221 is merged, so I rebased this pull request on top of master and updated the description.

pacak commented:
If there's no problems with the code - this pull request needs FCP

https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-03-28/near/430107731

jieyouxu commented:
@petrochenkov since you have initially reviewed this, would you like to start an FCP? (AFAIK this needs T-compiler FCP).

bjorn3 commented:
Looks like the cg_clif side got lost. Other than that LGTM.

pacak commented:
> Looks like the cg_clif side got lost. Other than that LGTM.

Hmm... Yea, looks like `compiler/rustc_codegen_cranelift/src/driver/aot.rs` contains a copy-pasted bit from `rustc_codegen_ssa`. Are there any plans to unify those two? If not - I can add the same logic for cranelift as well...

bjorn3 commented:
> Are there any plans to unify those two?

At some point in the future once cg_ssa is flexible enough for it. Will likely take a while.

pacak commented:
> Will likely take a while.

Okay, I'll add a copy of the same logic I added to cg_ssa then, it will only generate artifact notifications when user specifically asked to emit this type of artifacts.

pacak commented:
Seem to work for cranelift both single and multiple codegen units.

```
/path/to/rust/build/x86_64-unknown-linux-gnu/stage3/bin/rustc 
    -Zcodegen-backend=cranelift src/lib.rs 
    --emit obj,asm,llvm-ir,llvm-bc,mir --json artifacts --error-format=json
...
{"$message_type":"artifact","artifact":"lib.mir","emit":"mir"}
{"$message_type":"artifact","artifact":"lib.o","emit":"obj"}
...
```

```
{"$message_type":"artifact","artifact":"lib.lib.fc057f2d5d964f9f-cgu.0.rcgu.o","emit":"obj"}
{"$message_type":"artifact","artifact":"lib.lib.fc057f2d5d964f9f-cgu.1.rcgu.o","emit":"obj"}
{"$message_type":"artifact","artifact":"lib.lib.fc057f2d5d964f9f-cgu.2.rcgu.o","emit":"obj"}
```

bjorn3 reviewed (approved):
Minor suggestion about the method name. The actual change looks solid to me. r=me once FCP has finished.

pacak commented on compiler/rustc_codegen_ssa/src/lib.rs:
```diff
@@ -113,6 +113,27 @@ pub struct CompiledModule {
 
+impl CompiledModule {
+    /// Call `emit` function with every artifact type currently compiled
+    pub fn traverse_artifacts_with<F>(&self, mut emit: F)
```
Renamed.

pacak commented:
@petrochenkov or anyone, any progress with FCP? It's been sitting for a week...

jieyouxu commented:
> @petrochenkov or anyone, any progress with FCP? It's been sitting for a week...

I'll send a message in the T-compiler zulip thread to have a T-compiler team member start an FCP

wesleywiser commented:
@rfcbot fcp merge

pacak commented on compiler/rustc_codegen_ssa/src/lib.rs:
```diff
@@ -113,6 +113,27 @@ pub struct CompiledModule {
 
+impl CompiledModule {
+    /// Call `emit` function with every artifact type currently compiled
+    pub fn for_each_output<F>(&self, mut emit: F)
```
Hmmm... I guess this is not enough. Let me make a proper fix and rebase while at it.

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit c8390cdbfaccc55aa374222e39cdf9092a1e5ff6 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 c8390cdbfaccc55aa374222e39cdf9092a1e5ff6 -->
<!-- homu: {"type":"Approved","sha":"c8390cdbfaccc55aa374222e39cdf9092a1e5ff6","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c8390cdbfaccc55aa374222e39cdf9092a1e5ff6 with merge 1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0...
<!-- homu: {"type":"BuildStarted","head_sha":"c8390cdbfaccc55aa374222e39cdf9092a1e5ff6","merge_sha":"1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9358943218/job/25764971408)
Approved by: bjorn3
Pushing 1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9358943218/job/25764971408"},"merge_sha":"1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0"} -->

rust-timer commented:
Finished benchmarking commit (1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0): [comparison URL](https://perf.rust-lang.org/compare.html?start=7c52d2db6348b038276198e88a835125849f322e&end=1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=7c52d2db6348b038276198e88a835125849f322e&end=1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-0.9%, -0.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.2%, -1.2%] | 1     |
| All ❌✅ (primary)                 | -0.9% | [-0.9%, -0.9%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7c52d2db6348b038276198e88a835125849f322e&end=1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0&stat=max-rss)
<details>
<summary>Results (secondary 3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.2% | [2.9%, 3.5%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7c52d2db6348b038276198e88a835125849f322e&end=1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=7c52d2db6348b038276198e88a835125849f322e&end=1689a5a531f1fe404944ed8c3ac6cb85a2cff7e0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 671.059s -> 673.132s (0.31%)
**Artifact size**: 319.05 MiB -> 318.94 MiB (-0.03%)
